lib/reek/spec/should_reek_of.rb

```diff
@@ -17,22 +23,42 @@ def initialize(smell_category,
 
       def matches?(actual)
         self.examiner = Examiner.new(actual, configuration: configuration)
-        self.all_smells = examiner.smells
-        all_smells.any? { |warning| warning.matches?(smell_category, smell_details) }
+
+        set_failure_messages_for_smell_type && return unless matching_smell_types?
+        set_failure_messages_for_smell_details && return unless matching_smell_details?
+
+        true
```

Comment:
But none of your suggestions above deal with:

````This code gives the impression that both validations are independent from each other. But they are not. The matching_smell_details? one is dependent on the matching_smell_types? one because it can't stand on it's own.````

Which is why I still believe that my initial solution is better.

How about a compromise with something like this:

```Ruby
def matches?(actual)
  if matching_smell_types?
    if matching_smell_details?
      return true
    else
      set_failure_messages_for_smell_details
    end
  else
    set_failure_messages_for_smell_types
  end
  false
end
```

WDYT?